Remove attributes and method of submodules in frozen module

bzinodev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34787 Remove attributes and method of submodules in frozen module**

This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

Differential Revision: [D21004990](https://our.internmc.facebook.com/intern/diff/D21004990)

Commit:
Remove attributes and method of submodules in frozen module

This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

Commit:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

Commit:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

Commit:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

Commit:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

Commit:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

Commit:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

Commit:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -218,10 +190,13 @@ class AttributePropagator {
+      attr = std::move(tuple);
+
     } else if (attr.isList()) {
       c10::List<IValue> elems = std::move(attr).toList();
```
When we iterate over the dict here we're iterating as:
```
      for (const auto& pair : dict) {
        auto val = pair.value();
        val = overrideGradient(val);
      }
```

val here is a const pointer right so what does setting `val = overrideGradient(val);` do?

in general i dont really understand how this function works. with `Tensor` we are constructing a new tensor and returning it, list in place modifying, tuple and dict constructing or in-place modifying.

bzinodev replied:
Added std::move semantic so it is not in-place

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -318,55 +291,151 @@ class AttributePropagator {
   }
 
-  // cleanupFrozenModule function cleans up the Frozen module. it performs the
+  bool moduleEscapes(Module& subModule, std::shared_ptr<Graph>& graph) {
```
This doesn't check nested types. I think you want to recurse through the contained types of the graph outputs. Also I believe it is an invariant that jit graph should only ever have one output.

bzinodev replied:
it does not hurt to check every output of the graph. I rather check every output?

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -318,55 +291,151 @@ class AttributePropagator {
-  // cleanupFrozenModule function cleans up the Frozen module. it performs the
+  bool moduleEscapes(Module& subModule, std::shared_ptr<Graph>& graph) {
+    for (auto& output : graph->outputs()) {
+      if (output->type() == subModule.type()) {
```
This is checking pointer equality, I believe you want `*output->type() == *subModule.type()`, or `subModule.type()->isSubtypeOf(output->type())`

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -318,55 +291,151 @@ class AttributePropagator {
+
+  // Prepraring for clean up phase. At this point, record all  subModules that
+  // contains mutable attributes.
+  void recordReferencedAttrs(std::shared_ptr<Graph>& graph) {
```
How has this function changed? it's hard to tell because the code got reorganized.

bzinodev replied:
I had to change it. Because initial implementation checks only toplevel module. Now it is extended to scan submodules.

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -318,55 +291,151 @@ class AttributePropagator {
+      }
+      for (auto& fn : type->methods()) {
+        auto& name = fn->name();
+        if ("forward" == name && type == module_.type())
```
In this case you may have the same type ptr but probably safer to do `*type == module_.type()`

eellison reviewed (commented):
A few comments. @suo will all modules of the same class type have the same pointer ? EDIT: he says yes, this is true for class types / module types.

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -318,55 +291,151 @@ class AttributePropagator {
       preservedScalarAttrs_;
 
+  // Contains the attribute slots that need to be preserved for each ClassType.
+  std::unordered_map<ClassTypePtr, std::unordered_set<size_t>> attrsToKeep_;
```
I was wondering about it. My initial test concluded yes they are the same but I agree TypePtr is safer. I will make the change

bzinodev commented on test/jit/test_freezing.py:
```diff
@@ -117,9 +117,182 @@ def forward(self, x):
+        input = torch.randn(2, 2)
+        output_s = m.forward(input)
+        mf = torch._C._freeze_module(m._c)
+        self.assertTrue(mf.hasattr('sub1'))
```
Ok! good feedback

bzinodev commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -79,11 +76,8 @@ class AttributePropagator {
     }
 
     for (auto& moduleName : names_) {
-      auto it = preservedAttrs_.find(attrModule._ivalue());
```
formatting?

bzinodev committed:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -316,54 +291,151 @@ class AttributePropagator {
+      }
+      for (auto& fn : type->methods()) {
+        auto& name = fn->name();
+        if ("forward" == name && *type == *module_.type())
```
Why do we need to special case `forward` here ?

eellison reviewed (approved):
LGTM

bzinodev committed:
Update on "Remove attributes and method of submodules in frozen module"


This is a follow up patch of freezing of TorchScript modules. This patch
enables removal of constant attributes and unused method in submodules.
The clean up logic is generalized to handle  attributes that share their class
type.

[ghstack-poisoned]
